### ISSUE TITLE
iOS: File uploads cause high ram usage

### ISSUE THREAD

**Issue by dlewis23** [tidev/titanium_mobile#13320]
### I have searched and made sure there are no existing issues for the issue I am filing

- [X] I have searched the existing issues

### Description

When doing a file upload using Ti on iOS the file is placed into ram instead of being uploaded from the disk, this is fine for a small file but when you do a larger file 40 MB+ or multiple files this can become an issue. A file that 40 MB in size on disk uses over 100 MB of ram when it's placed there to do the upload. 

The file is cleared from ram after the upload is complete, unless you are doing multiple uploads where the files can remain till all uploads are complete, but will not always be removed. This can cause an app to be using well over 1 GB of ram when it should be less than 100 MB. 

**Sample**

A 100MB file with the sample code below will use 400 MB of ram when uploading.

`var xhr = null;
    
   var win = Ti.UI.createWindow({
     backgroundColor: '#fff',
     fullscreen: false, navBarHidden: true, exitOnClose: true, theme: "Theme.materialTheme", orientationModes: [Titanium.UI.PORTRAIT]
   });
    
   var runTime = Ti.UI.createLabel({
     text: 'Run Time...',
     top: 50,
     left: 10,
     color: '#000'
   });
    
   var uploadSpeed = Ti.UI.createLabel({
     text: '0.00 Mbps',
     top: 90,
     left: 10,
     color: '#000'
   });
    
   var buttonStart = Ti.UI.createButton({ title: "START", width: 100, height: 40, top: '45%', left: '10%' });
   var buttonStop = Ti.UI.createButton({ title: "STOP", width: 100, height: 40, top: '45%', right: '10%' });
    
   var uploadfile = Ti.Filesystem.getFile(Ti.Filesystem.getResourcesDirectory(), "100MB.bin").read();
    
   buttonStart.addEventListener("click", function() {
    
     console.log('Starting Upload');
    
     var startTime = new Date().getTime();
     var totalPercent = 0;
    
     xhr = Titanium.Network.createHTTPClient();
     xhr.onsendstream = function(e){ 
         console.log(e.progress);
    
         totalPercent = (e.progress).toFixed(2);
         var currentTime = new Date().getTime();
         var timeDiff = ((currentTime-startTime)/1000).toFixed(2);
         var totalDownloaded = (102400 * totalPercent);
         var currentKBPS = (((((totalDownloaded * 1024) * 8.192) / timeDiff) / 1024 / 1024)).toFixed(2);
    
         runTime.text = timeDiff;
         uploadSpeed.text = currentKBPS + ' Mbps';
     };
    
     xhr.onload = function(e) {
    
     };
     xhr.open('POST','http://sfo.veeapps.com/upload.php');
     xhr.send({media: uploadfile });
     
   });
    
    
   buttonStop.addEventListener("click", function() {
     xhr.abort();
     xhr = null;
   });
    
   win.add(runTime);
   win.add(uploadSpeed);
   win.add(buttonStart);
   win.add(buttonStop);
   win.open();
   `

### Solution

This solution from Stack has been suggested: https://stackoverflow.com/questions/12626748/what-is-the-correct-way-of-sending-large-file-through-http-post-without-loading/12627138#12627138

I believe we need the option to upload from disk not from ram.

### Platforms

iOS

**Comment by hansemannn** [comment]
The solution is unfortunately very outdated, since the `NSURLConnection` has been deprecated a long time ago in favor of the `NSURLSession` API. We spent huge amounts of work into optimizing the memory footprint of POST requests in the past, but it seems like this is still an issue. I am not sure where to start from tbh.

**Comment by dlewis23** [comment]
Ok. Shoot. This is so odd because TI is the only one this happens with. If you do a POST in react native for example it uploads it from disk the file does not go into ram. 

There has to be a way to make a POST with out it adding the file into ram, if you want to upload a file that is like a 1 gig video you can't put that into ram on a iOS device it would crash the app. 

I will continue to look around and see if I come across anything.

**Comment by hansemannn** [comment]
I agree, it's totally fixable, no worries! Would you mind trying https://github.com/tidev/ti.urlsession to upload the files? Its a more lightweight version for networking. And if possible, pass a file path instead a blob, because a blob needs to read the file into memory.

**Comment by dlewis23** [comment]
So I tired to use ti.urlsession and ended up just getting the error below. 

`[ERROR] Error: Upload tasks from NSData are not supported in background sessions.`

**Comment by hansemannn** [comment]
I have to check that. Do you have a single app js example? If possible, even with a sample
file?

**Comment by hansemannn** [comment]
Looking at https://stackoverflow.com/questions/25660363/upload-tasks-from-nsdata-are-not-supported-in-background-sessions, it seems like you should pass the path, not the blob (as mentioned earlier as well).

**Comment by dlewis23** [comment]
I saw that too and I did pass a path to the file like below. With the uploadData file below. 

`

	sessionConfig = URLSession.createSessionConfiguration({
		identifier: 'com.test.test2'
	});

	session = URLSession.createSession({
		configuration: sessionConfig
	});

	var uploadData = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'upload.zip').read();

	session.uploadTask({
		url: 'http://sfo.veeapps.com/upload.php',
		data: uploadData
	});
`

**Comment by hansemannn** [comment]
Yes, and instead of
> var uploadData = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'upload.zip').read();
do
> var dataPath = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, 'upload.zip').nativePath
and pass it as the "data" parameter.

**Comment by dlewis23** [comment]
Super long time. But I have been back on this one this week... 

While ti.urlsession does not put the file into ram when it is uploaded, it just doesn't seem to upload very fast or consistent when compared to a normal XHR.
